### PR TITLE
SAML login usernames do not allow \h, and instead allow the usage of \\h with ensured JWT consistency

### PR DESCRIPTION
**Pull request by nancyjlau**
### Description
Addressed an issue where SAML login would fail for usernames containing `\h`. While usernames with `\h` remain unsupported, this fix ensures that `\\h` can be used as a consistent alternative. Usernames with an even number of backslashes followed by `h` will always be treated as having just `\\h`.

**Category:** Bug fix

**Reason for Change:** 
Usernames containing `\h` were found to cause invalid JWT token generation, as detailed in [Issue 2531](https://github.com/opensearch-project/security/issues/2531). However, `\\h` produces valid JWT tokens. Due to inconsistent handling of backslashes (owing to their use as escape characters in string literals), it's crucial to establish a consistent approach. Hence, any even number of backslashes before `h` will now be treated as `\\h`.

**Behavioral Change:**
- **Previous Behavior:** SAML login fails when the username contains `\h`.
- **New Behavior:** SAML login succeeds for usernames with `\\h` and fails for `\h`.

### Issues Resolved
- Resolves https://github.com/opensearch-project/security/issues/2531

### Testing
A test `testJwtWithBackslashHInUsername` was added in `src/test/java/com/amazon/dlic/auth/http/saml/HTTPSamlAuthenticatorTest.java` to validate the behavior with various username patterns. This test iterates over an array of test usernames, ranging from "username\hexample" to "username\\\\\\\\hexample", to ensure that only the `\\h` format is valid.

In addition to the automated tests, manual testing was conducted where tokens were created with usernames containing `\h` (which consistently failed) and `\\h` (which succeeded). The validity of these tokens was further confirmed using the [jwt.io site created by Okta](https://jwt.io/).

### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Commits are signed per the DCO using `--signoff`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. More information on the Developer Certificate of Origin and signing off on commits can be found [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/com/amazon/dlic/auth/http/saml/AuthTokenProcessorHandler.java [hunk 1]
```diff
@@ -310,12 +312,12 @@ private String createJwt(SamlResponse samlResponse) throws Exception {
 
         if (token_log.isDebugEnabled()) {
             token_log.debug(
-                "Created JWT: "
-                    + encodedJwt
-                    + "\n"
-                    + jsonMapReaderWriter.toJson(jwt.getJwsHeaders())
-                    + "\n"
-                    + JwtUtils.claimsToJson(jwt.getClaims())
+                    "Created JWT: "
```
Comment (DarshitChanpura) [inline src/main/java/com/amazon/dlic/auth/http/saml/AuthTokenProcessorHandler.java:315]: nit: seems like unwanted indentation changes.

File: src/test/java/com/amazon/dlic/auth/http/saml/HTTPSamlAuthenticatorTest.java [hunk 2]
```diff
@@ -906,6 +906,52 @@ private RestRequest buildTokenExchangeRestRequest(
             .withHeaders(ImmutableMap.of("Content-Type", "application/json"))
             .build();
     }
+    @Test
+    public void testJwtWithBackslashHInUsername() throws Exception {
+        String[] testUsernames = {
```
Comment (DarshitChanpura) [inline src/test/java/com/amazon/dlic/auth/http/saml/HTTPSamlAuthenticatorTest.java:911]: Could you add a test with failure scenario as well? `\h`

File: src/test/java/com/amazon/dlic/auth/http/saml/HTTPSamlAuthenticatorTest.java [hunk 3]
```diff
@@ -906,6 +906,52 @@ private RestRequest buildTokenExchangeRestRequest(
...
+                    }
+            );
+
+            String authorization = (String) response.get("authorization");
+            Assert.assertNotNull("Expected authorization attribute in JSON for input: " + testUsername, authorization);
+
+            JwsJwtCompactConsumer jwtConsumer = new JwsJwtCompactConsumer(authorization.replaceAll("\\s*bearer\\s*", ""));
+            JwtToken jwt = jwtConsumer.getJwtToken();
+
+            Assert.assertEquals("username\\hexample", jwt.getClaim("sub"));
```
Comment (DarshitChanpura) [inline src/test/java/com/amazon/dlic/auth/http/saml/HTTPSamlAuthenticatorTest.java:952]: nit: Can you replace with assertThat from hamcrest? There is an effort to  change assertEquals to assertThat (to be more human readable).
```java
assertThat(jwt.getClaim("sub"), is(equalTo("username\\hexample")));
```

**Review by DarshitChanpura — COMMENTED** [review]
Thank you @nancyjlau for this contribution and details in the PR description! I have left some nit comments and one request to add a test for `\h`

File: src/main/java/com/amazon/dlic/auth/http/saml/AuthTokenProcessorHandler.java (hunk 1 above)
Comment (nancyjlau) [reply src/main/java/com/amazon/dlic/auth/http/saml/AuthTokenProcessorHandler.java:315]: Oops, this part was an accident, will fix asap

File: src/test/java/com/amazon/dlic/auth/http/saml/HTTPSamlAuthenticatorTest.java (hunk 2 above)
Comment (nancyjlau) [reply src/test/java/com/amazon/dlic/auth/http/saml/HTTPSamlAuthenticatorTest.java:911]: Will do!

**Comment by DarshitChanpura** [comment]
@nancyjlau Would you need any help getting this across the finish line?

**Comment by nancyjlau** [comment]
Hi, sorry I had gotten distracted by an exam last week, will submit changes for this today!

**Comment by cwperks** [comment]
This really is specific to `\h` and it doesn't affect other special characters like `\t` or `\n` from some local testing.

I tried:

- `user\hexample`
- `user\texample`
- `user\nexample`

and only the one where the subject contains `\h` is where the error appeared. 

As @MaciejMierzwa mentioned [here](https://github.com/opensearch-project/security/issues/2531#issuecomment-1755556797) it looks like either Java or the JWT Library (CXF or Nimbus) are not handling `\h` as they are `\n` and `\t`. I'd like to understand why `\h` is special compared to the other chars.

**Comment by DarshitChanpura** [comment]
@nancyjlau Would you need any help wrapping this PR up? I'm marking this as draft meanwhile. Please mark it as ready for review once all changes are finalized and CI is green.

**Comment by nancyjlau** [comment]
uhhh sorry i have been having some git issues on trying to get CI to be green

**Comment by scrawfor99** [comment]
Hi @nancyjlau, you will need to fix the merge conflicts before the CI will be able to pass. Then we can see where things are at.

**Comment by DarshitChanpura** [comment]
Hi @nancyjlau Closing this as there has been no recent activity. Feel free to @ if this changes.

**Pull request closed without merging**